src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs

```diff
@@ -534,44 +526,88 @@ private static async Task WriteDataAsync(Stream archiveStream, Stream dataStream
             cancellationToken.ThrowIfCancellationRequested();
 
             await dataStream.CopyToAsync(archiveStream, cancellationToken).ConfigureAwait(false); // The data gets copied from the current position
+
             int paddingAfterData = TarHelpers.CalculatePadding(actualLength);
-            await archiveStream.WriteAsync(new byte[paddingAfterData], cancellationToken).ConfigureAwait(false);
+            if (paddingAfterData != 0)
+            {
```

Comment:
The assert that was added in the sync version:
```suggestion
            {
                Debug.Assert(paddingAfterData <= TarHelpers.RecordSize);

```

Reply:
I can add it in a separate PR later this week.

Reply:
Because the sync version stackallocs.  The async version doesn't (can't).